eventually-new/src/aggregate.rs

```diff
@@ -0,0 +1,375 @@
+//! Foundation traits for creating Domain abstractions
+//! using [the `Aggregate` pattern](https://martinfowler.com/bliki/DDD_Aggregate.html).
+
+use std::error::Error as StdError;
+use std::fmt::{Debug, Display};
+
+use async_trait::async_trait;
+
+use futures::stream::TryStreamExt;
+
+#[cfg(feature = "serde")]
+use serde::Serialize;
+
+use crate::eventstore::{EventStore, PersistedEvent, Select, Version};
+use crate::{Event, Events};
+
+/// A short extractor type for the Aggregate [`Id`].
+///
+/// [`Id`]: trait.Aggregate.html#associatedtype.Id
+pub type IdOf<A> = <A as Aggregate>::Id;
+
+/// An Aggregate manages a domain entity [`State`], acting as a _transaction boundary_.
+///
+/// It allows **state mutations** through the use of [`Command`]s, which the
+/// Aggregate instance handles and emits a number of Domain [`Event`]s.
+///
+/// [`Event`]: trait.Aggregate.html#associatedtype.Event
+/// [`State`]: trait.Aggregate.html#associatedtype.State
+/// [`Command`]: trait.Aggregate.html#associatedtype.Command
+#[async_trait]
+pub trait Aggregate: Send + Sync {
+    /// Aggregate identifier: this should represent an unique identifier to refer
+    /// to a unique Aggregate instance.
+    type Id: PartialEq + Send + Sync;
+
+    /// State of the Aggregate: this should represent the Domain Entity data structure.
+    type State: Default + Send + Sync;
+
+    /// Represents a specific, domain-related change to the Aggregate [`State`].
+    ///
+    /// [`State`]: trait.Aggregate.html#associatedtype.State
+    type DomainEvent: Send + Sync;
+
+    /// Commands are all the possible operations available on an Aggregate.
+    /// Use Commands to model business use-cases or [`State`] mutations.
+    ///
+    /// [`State`]: trait.Aggregate.html#associatedtype.State
+    type Command: AsRef<Self::Id> + Send + Sync;
+
+    /// Possible failures while [`apply`]ing [`Event`]s or handling [`Command`]s.
+    ///
+    /// [`apply`]: trait.Aggregate.html#method.apply
+    /// [`Event`]: trait.Aggregate.html#associatedtype.Event
+    /// [`Command`]: trait.Aggregate.html#associatedtype.Command
+    type HandleError: StdError + Send + Sync;
+
+    type ApplyError: StdError + Send + Sync;
+
+    /// Handles the requested [`Command`] and returns a list of [`Event`]s
+    /// to apply the [`State`] mutation based on the current representation of the State.
+    ///
+    /// [`Event`]: trait.Aggregate.html#associatedtype.Event
+    /// [`State`]: trait.Aggregate.html#associatedtype.State
+    /// [`Command`]: trait.Aggregate.html#associatedtype.Command
+    async fn handle(
+        &mut self,
+        state: &Self::State,
+        command: Self::Command,
+    ) -> Result<Option<Events<Self::DomainEvent>>, Self::HandleError>;
```

Comment:
What is the difference between None and empty vector?

Reply:
No difference, only saving a `Vec` allocation :smile:

Reply:
I'm aware of this "ambiguity", still wondering whether this is a good API design choice or not... What do you think?

Reply:
> No difference, only saving a Vec allocation

But as far as I remember `Vec::new()` doesn't allocate memory.

> What do you think?

I think I'd prefer unambiguous API. But I'm just passing by... :)

Reply:
I appreciate the feedback, the ambiguity is now resolved :smile:

Reply:
ah this was exactly my question in #157!